### ISSUE TITLE
CalDAV/CardDAV setup not possible anymore on Sailfish OS / Jolla devices

### ISSUE THREAD

**Issue by znerol** [nextcloud/server#1603]
### Steps to reproduce

On any Sailfish OS based device (e.g. Jolla Phone), add a CalDAV/CardDAV account and point the URL to a Nextcloud 10 server.
### Expected behaviour

Account created.
### Actual behaviour

An error message appears stating that something went wrong during account creation. This is due to the security hardening measures in `remote.php` (introduced in #50). The Sailfish OS account creation tool apparently checks the HTTP status and concludes that status 503 is non-recoverable.

PR #1601 resolves the problem, building on the work done in #1328.

**Comment by nickvergessen** [comment]
Unluckily we can not add an Exception for `Mozilla/5.0` since this is what everyone uses.
So I guess you have to do it by yourself.

**Comment by nickvergessen** [comment]
@LukasReschke any idea how to proceed with these kind of issues? Should we add a switch to disable security? 🙈 🙉 🙊

**Comment by LukasReschke** [comment]
No. As said in https://github.com/nextcloud/server/pull/1601#issuecomment-250927757 I'd require a traffic dump to see what is exactly happening there.

**Comment by LukasReschke** [comment]
Basically `\OC\AppFramework\Http\Request::passesStrictCookieCheck` should return `true` here. But why it doesn't is the interesting question. ;)

**Comment by znerol** [comment]
> I'd require a traffic dump to see what is exactly happening there.

I'm trying to come up with a reproducible setup (using the Nexcloud VM and the Emulator included in the Sailfish SDK), but I'm not there yet, i.e., everything which follows is pure speculation.

> since no cookies should be sent at all

That's actually what I was thinking also. Maybe the cookie storage is shared with the web-browser. That might spoil every subsequent API request.

Taking this thought a bit further I attempted to strip the cookie request header on the web-server level. But as far as I can tell `remote.php` requires some cookies to be present. At least that experiment was not successful and the API endpoint was not accessible for any device when no cookies at all were on the request.

Anyway, I think that any cookie directed to `remote.php` is superflous. Consequently one might argue that any `Set-Cookie` header an an API response is a bug.

**Comment by znerol** [comment]
Another thing we could try is to just issue a temporary redirect (i.e., HTTP status 302 or 307 instead of 503) when the same-site cookies need to be set in the client.

**Comment by znerol** [comment]
Regarding cookies on `remote.php`:

```
$ curl -s -D - -o /dev/null 'https://demo.nextcloud.com/remote.php/dav'
HTTP/1.1 401 Unauthorized
Cache-Control: no-store, no-cache, must-revalidate
Content-Length: 306
Content-Security-Policy: default-src 'none';script-src 'self' demo.nextcloud.com demo.nextcloud.services.stpdns.net stats.nextcloud.opensourcesoftware.rocks 'unsafe-inline' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob: stats.nextcloud.opensourcesoftware.rocks;font-src 'self';connect-src 'self';media-src 'self'
Content-Type: application/xml; charset=utf-8
Date: Tue, 04 Oct 2016 19:40:07 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Pragma: no-cache
Server: Apache
Set-Cookie: oci3kj7hqoh3=m508bkifvoq46rr8jkputcbvm5; path=/; HttpOnly
Set-Cookie: oc_sessionPassphrase=FuILAiDU9aBIodfMgn5jm0l9KX7qIezYUnWF1IBFeOQFNYBCKtcqPGlltxxCKf%2FxWOkPQXIQtc8w7hMcpjREVfYbs%2B6sM8oV2hOMnILao%2B0v2xBqYm0V%2Fx%2BxCTYlmZ%2Bu; path=/; secure; HttpOnly
Set-Cookie: nc_sameSiteCookielax=true; path=/; httponly;secure; expires=Fri, 31-Dec-2100 23:59:59 GMT; SameSite=lax
Set-Cookie: nc_sameSiteCookiestrict=true; path=/; httponly;secure; expires=Fri, 31-Dec-2100 23:59:59 GMT; SameSite=strict
Strict-Transport-Security: max-age=15768000; includeSubDomains; preload
Www-Authenticate: Basic realm="Nextcloud"
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Frame-Options: SAMEORIGIN
X-Permitted-Cross-Domain-Policies: none
X-Robots-Tag: none
X-Xss-Protection: 1; mode=block
```

Not sure, maybe this is due to the auto-login feature of the demo installation?

**Comment by LukasReschke** [comment]
> Anyway, I think that any cookie directed to remote.php is superflous. Consequently one might argue that any Set-Cookie header on an API response is a bug.

This is at the moment the expected behaviour since our sync clients rely on cookies for authentication. (since authenticating using Basic Auth is CPU expensive) - That may get changed with a Bearer Token in the future.

**Comment by znerol** [comment]
Playing with `cadaver` now in order to better understand the behavior. I use an stunnel directing 127.0.0.1:80 to demo.nextcloud.com:443 that makes it easy to capture the traffic.

```
$ cat <<'EOF' > stunnel-nextcloud-demo.conf
foreground = yes
pid =

[nextcloud-demo]
client = yes
accept = 127.0.0.1:80
connect = 176.9.48.186:443
```

```
sudo stunnel-nextcloud-demo.conf
```

Then point `demo.nextcloud.com` to `127.0.0.1` in `/etc/hosts` such that `cadaver` uses the proper `Host` HTTP header. Then capture using `tcpdump` on the local loopback interface on port 80.

Interestingly `cadaver` never sends any cookies. That in turn leads to nextcloud generating a new session on every request. It might be possible to fix that behavior by disabling session authentication when there is an `Authorization` header on the request.

**Comment by blizzz** [comment]
@znerol got my Jolla C (SF 2.0.2.51) on the weekend and I did not have any issues adding my Nc 10 Cal- and Carddav. Not aware that anything changed inbetween?

**Comment by MorrisJobke** [comment]
Will be fixed with #8123 and there is even a PR available which could come to Nextcloud 14. Thus closing this as a duplicate of it.
